notebooks/advanced/Time_Series_Forecasting_with_Unstructured_Data_and_Amazon_SageMaker_Neural_Topic_Model/2_NTM.ipynb

```diff
@@ -0,0 +1,838 @@
+    "print('vocab size:', len(vocab_list))\n",
+    "\n",
+    "# Retain the index of the vectors\n",
+    "idx = np.arange(vectors.shape[0])\n",
+    "print('Done. Time elapsed: {:.2f}s'.format(time.time() - start_time))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Type cast the vectors into a sparse array\n",
+    "import scipy.sparse as sparse\n",
+    "vectors = sparse.csr_matrix(vectors, dtype=np.float32)\n",
+    "topic_vectors = sparse.csr_matrix(topic_vectors, dtype=np.float32)\n",
+    "print(type(vectors), vectors.dtype)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Training a Neural Topic Model\n",
+    "\n",
+    "To extract \"text vectors\", we take the Topics and convert each topic into a 20 (NUM_TOPICS) dimensional \"Topic\" vector. This can be viewed as an effective lower dimensional embedding of all the text in the corpus into some predefined topics. Each Topic will have a representation as a vector, and related topics will have a related vector representation. Assuming that there is some correlation between topics from one day to the next., i.e, the top topics don't change very frequently on a daily basis, we can represent all the text in the dataset as a collection of 20 topics. Feel free to experiment with changing the number of topics to extract by modifying the NUM_TOPICS parameter.\n",
+    "\n",
+    "Each topic can be viewed as distinct from every other topic, allowing us to then input it as a separate timeseries into DeepAR for model training."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#IAM Roles:- In order to train a model on SageMaker, SageMaker needs to assume an IAM role to be able to access data in S3.\n",
+    "## If you are running in a SageMaker notebook environment, a role has already been created when you provisioned the notebook instance.\n",
+    "\n",
+    "# To use that role simply run the following cell. If you get an exception, create a SageMaker IAM role first using this link: https://docs.aws.amazon.com/glue/latest/dg/create-an-iam-role-sagemaker-notebook.html\n",
+    "# Make sure that along with the AmazonSageMakerFullAccess Policy you also attach a policy to provide access to your specific S3 bucket\n",
+    "# This could be easily done by providing AmazonS3FullAccess but this is **not** a recommended security best practice.\n",
+    "# then input that role in the text widget below.\n",
+    "\n",
+    "try:\n",
+    "    from sagemaker import get_execution_role\n",
+    "    role = get_execution_role()\n",
+    "except Exception as e:\n",
+    "    role_widget = util.create_text_widget( \"IAM Role Name\", \"Enter the IAM role needed for SageMaker to access training containers\" )\n",
+    "    role = role_widget.value\n",
+    "    assert role, \"role not set.\""
```

Comment:
In local Jupyter environment, customers always get "AssertionError: role not set." when this cell is executed first time. 

Like other places, split the latter 2 lines as separate cell.

Alternatively, you can use Python's input() function. ( I recently found that input() is easier to accept customer's text input interactively than ipywidget's text widget.